### ISSUE TITLE
info, fixed ffmpeg bug to start frm/teema 4 hours earlier

### ISSUE THREAD

**Issue by gunnar12** [aajanki/yle-dl#108]
in hls.c, its kindo done already but always does nothing.
copy from my ffmpeg-quick-wiki-ticket
function select_cur_seg_no, at line abt 1500 to 1600 hls.c
-----
 just fixed the funny live_start_index bug.
At least fix for the server and stream I am interested in.
Turns out like this in my case (10s segments)

    pls->start_seq_no, whatever segment is streamed "now" from stream server
    pls->n_segments, the max OLD segments the server keeps available, 1438,FOUR HOURS
    c->live_start_index, where ffmpeg puts the command argument I give it, 1400 

Note, pls->n_segments is given as a positive number although it is kindo negative, how max cmaller one can make the segmentcounter. (the thought-bug?)

THat is, a funny thing, shld i use PLUS 1400 or MINUS 1400?
Note, when asking for old, smaller number segments!

(seems the existing code did the same thought-bug in both cases?? always did nothing?)

So I made it accept both but always GO BACKWARDS (earlier), both in time and segments

    /* index If this is a LIVE stream, start live_start_index segments from the
    start or end */ 

        printf("\n start_seq_no %d n_segments %d live_start_index %d \n",pls->start_seq_no,pls->n_segments,c->live_start_index) ;

        if (c->live_start_index < 0) /* user gives neg number, 5th grade -/+ math*/
            /* take max of two negative numbers, hehe*/
            /* 93abcd, - 1438(4 hours back), -1400*/
        seq_gunnar = pls->start_seq_no + FFMAX((- pls->n_segments), c->live_start_index);
    else /* user gives pos number, kindo simpler, just subtract the smaller*/
        /* 1400, 1438 */
    seq_gunnar = pls->start_seq_no - FFMIN(c->live_start_index, pls->n_segments);

        printf("\n start_seq_no %d n_segments %d live_start_index %d gunnar %d\n",pls->start_seq_no,pls->n_segments,c->live_start_index,seq_gunnar) ;

        return seq_gunnar;
----------------------
bottom line, the negative of a cpl of negative or not negative numbers is funny

**Comment by gunnar12** [comment]
this "start form xxxx segments ago" command for ffmpeg is
-live_start_index 1400

had already added same "feature" to adobe.hds.

I know,i know, yle-dl is not really 
for downloading YLE live streams, but could be??

ffmpeg also seems to need some "robustness fixes",  at least the win32 version (crosscompiled on ubuntu)
For me the downloaf gets "stuck" now and then, and cant resume.
Found a brute force fix, use microsoft TcpView to find and disconnect the stuck connection.
Ffmpeg continues with next segment but the segment just under way gets truncated,

Planning to fix ffmpeg to do same as adobehds, first store the segments to a temp dir and only when fully received start using them.
This wld make resume easy, delete the video file but not the temporary segment files (except for last ones), restart download, which then uses old stored temp segments until it has to start downloading.

**Comment by gunnar12** [comment]
win32 typical 51min ffmeg bug, abd avi vs mp4 output.

If using avi output one can start watching (wth mplayer) immediatly.
For mp4 one has to wait untill full video downloaded, bc only then are video-index written to the end of the mp4 file.
AVI:
Mplayer makes its own internal index when avi file opened,
(however, one cannot navigate frwd,bkwrds beyond the size of the file when mplayer started. But mplayer continues beyond that point. Need to open video file again, wait, and skip forward)
 
MP4, needs to wait until done, index written to end of file (mplayer baddy)

BUTBUTBUT, even with ffmpeg-mplayer, the playback cant get past 51 minutes.
Maybe only on 32 bit version because it seems to be the limit of some 32-bit thing, kindo 3200 seconds and internally done in us, giving appr 3.200.000. (avi standard prblm too?)
When ffmpeg is done, exits, it fixes this thing mplayer cant get past, but then it is as bad as mp4.

Planning to have yle-dl and ffmpeg to output both avi and mp4 files (containers), then delete the avi.


PS and finally, the old "non-mono packet-data stamps" of ffmpeg, always there on esp live streams..
(fix by first storing segments to temp dir?)

For YLE live it is really silly, fixes back and forth a;most forever when started..typ bug..

PPS one can ask, why did yle-dl buddy up with ffmpeg?? hehe

**Comment by aajanki** [comment]
If you have questions about ffmpeg or patches to ffmpeg you should send them to the ffmpeg development mailing list, because I'm not familiar with ffmpeg source code: https://ffmpeg.org/developer.html#Submitting-patches

BTW, yle-dl can download live broadcasts using AdobeHDS, for example: `yle-dl http://areena.yle.fi/tv/suorat/yle-teema-fem`
